components/browser/session/src/main/java/mozilla/components/browser/session/SessionManager.kt

```diff
@@ -80,6 +80,13 @@ class SessionManager(
 
             session.engineSessionHolder.engineSession = null
             session.engineSessionHolder.engineObserver = null
+
+            // Note: We could consider not clearing the engine session state here. Instead we could
+            // try to actively set it here by calling save() on the engine session (if we have one).
+            // That way adding the same Session again could restore the previous state automatically
+            // (although we would need to make sure we do not override it with null in link()).
+            // Clearing the engine session state would be left to the garbage collector whenever the
+            // session itself gets collected.
```

Comment:
Interesting, yes. I do also prefer the current fix, as it's a bit easier to understand i.e. we only save the engine session state and unlink if we have an engine session, and eagerly remove the old state in unlink, but both seem fine.

Reply:
Yeah, always keeping (or actively saving) the state may make it easier for the current "undo" workaround of re-adding a `Session`. But I also would like to not make this change for now if not really needed. :)